# Proyecto: "Déjame escuchar música" (TripleTen)

In [53]:
import pandas as pd# Importa pandas


In [54]:
df=pd.read_csv("music_project_en.csv")# Lee el archivo y almacénalo en df


In [55]:
df.head(10)# Obtén las 10 primeras filas de la tabla df


,userID,Track,artist,genre,City,time,Day
0,FFB692EC,Kamigata To Boots,The Mass Missile,rock,Shelbyville,20:28:33,Wednesday
1,55204538,Delayed Because of Accident,Andreas Rönnberg,rock,Springfield,14:07:09,Friday
2,20EC38,Funiculì funiculà,Mario Lanza,pop,Shelbyville,20:58:07,Wednesday
3,A3DD03C9,Dragons in the Sunset,Fire + Ice,folk,Shelbyville,08:37:09,Monday
4,E2DC1FAE,Soul People,Space Echo,dance,Springfield,08:34:34,Monday
5,842029A1,Chains,Obladaet,rusrap,Shelbyville,13:09:41,Friday
6,4CB90AA5,True,Roman Messer,dance,Springfield,13:00:07,Wednesday
7,F03E1C1F,Feeling This Way,Polina Griffith,dance,Springfield,20:47:49,Wednesday
8,8FA1D3BE,L’estate,Julia Dalia,ruspop,Springfield,09:17:40,Friday
9,E772D5C0,Pessimist,NaN,dance,Shelbyville,21:20:49,Wednesday


In [56]:
df.info()# Obtén la información general sobre nuestros datos


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


Hallazgos en el df

1.-En las filas hay registro del momento y lugar en el que un usuario escuchó una canción específica de un artista específico.



2.-En df.info() dice:
RangeIndex: 65079
entries, 0 to 65078

3.-Algunas columnas tienen más de mil valores ausentes.
Todas las columnas dicen "Dtype object". No es posible saber si hay valores incorrectos.

# Limpieza de datos en el df

In [57]:
# Mostrar los nombres de las columnas
df.columns

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')

In [58]:
# Bucle que itera sobre los encabezados y los pone todos en minúsculas
column_names_lowered=[]
for header in df.columns:
    lower_name=header.lower()
    column_names_lowered.append(lower_name)
df.columns=column_names_lowered
df.columns

Index(['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day'], dtype='object')

In [59]:
# Bucle que itera sobre los encabezados y elimina los espacios
column_no_space=[]
for header in df.columns:
    no_space_name=header.strip()
    column_no_space.append(no_space_name)
df.columns=column_no_space
df.columns


Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

In [60]:
# Cambia el nombre de la columna "userid" para seguir la regla de snake_case
df.rename(columns={"userid": "user_id"}, inplace=True)

#Comprobar los resultados
df.columns



Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

In [61]:
#Calcular el número de valores ausentes
df.isna().sum()


user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64

In [62]:
#Bucle en los encabezados para reemplazar los valores ausentes con 'unknown'
columns_to_replace=['track', 'artist', 'genre']
for column in columns_to_replace:
    df[column] = df[column].fillna('unknown')

#Contar los valores ausentes
df.isna().sum()

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64

In [63]:
#Contar los duplicados explícitos
df.duplicated().sum()

3826

In [64]:
#Eliminar duplicados explícitos
df = df.drop_duplicates()


In [65]:
#Comprobar de nuevo si hay duplicados en el df
df.duplicated().sum()

0

In [66]:
#Inspeccionar los nombres de géneros únicos
df_genres=df["genre"].unique()
print(sorted(df_genres))



['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

In [67]:
#Definir función para reemplazar los duplicados implícitos
def replace_wrong_values(df,column,wrong_values,correct_value):
    for col in df:
        df[col].replace(wrong_values, correct_value)



In [68]:
#Usar la función paea eliminar los duplicados implícitos
replace_wrong_values(df,["genre"],["hip","hop","hip-hop"],"hiphop")


In [69]:
#Comprobar de nuevo los duplicados implícitos
df_genres=df["genre"].unique()
print(sorted(df_genres))


['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

## Análisis

In [70]:
#Contar las canciones reproducidas en cada ciudad
df.groupby('city')['track'].count()


city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64

# Inferencia:

En Springfield se escuchan más del doble de canciones que en Shelbyville.

Esto podría deberse a la legalización del incesto (entre primos), lo cual, produciría una proliferación de defectos genéticos que convergerían en una tasa de mortalidad infantil y neonatal más alta de lo normal. Así, una menor cantidad de población provoca una menor cantidad de usuarios.

Otra razón podría ser la diferencia de tamaño de ambas ciudades, de la mano con la densidad de población (la cual no se ha específicado en los datos).

Otra razón pudiera deberse meramente a diferencias culturales en cuanto al uso del servicio de streaming.

In [71]:
#Calcular las canciones reproducidas lunes y viernes
df_by_days=df.groupby('day')['track'].count()
print(df_by_days.loc[["Monday","Friday"]])



day
Monday    21354
Friday    21840
Name: track, dtype: int64


El viernes se reprodujeron más canciones que el lunes.

In [72]:
#Definir la función number_tracks() con dos parámetros: day= y city=
def number_tracks(day,city):

    #Almacenar las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=
    filter_by_day=df[df['day'] == day]
    #Filtrar las filas donde el valor en la columna 'city' es igual al parámetro city=
    filter_by_city=filter_by_day[filter_by_day['city'] == city]
    #Extraer la columna 'user_id' de la tabla filtrada y aplica el método count()
    result=filter_by_city["user_id"].count()
    #Devuelver el número de valores de la columna 'user_id'
    return result

In [73]:
#Llamar a la función para determinar el número de canciones reproducidas en Springfield el lunes
print(number_tracks("Monday","Springfield"))


15740


In [74]:
#Determinar el número de canciones reproducidas en Shelbyville el lunes
print(number_tracks("Monday","Shelbyville"))



5614


In [75]:
#El número de canciones reproducidas en Springfield el viernes
print(number_tracks("Friday","Springfield"))



15945


In [76]:
# El número de canciones reproducidas en Shelbyville el viernes
print(number_tracks("Friday","Shelbyville"))


5895


# Conclusiones <a id='end'></a>

Lunes-Springfield=15,740  
Lunes-Shelbyville=5,614  
Viernes-Springfield=15,945  
Viernes-Shelbyville=5,895

En ambas ciudades, el viernes se ecuchó más música que el lunes.  
Springfield, al tener más usuarios que Shelbyville, arrojará una cantidad mayor de reproducciones que Shelbyville.

La fuente original, el archivo scv, tenía una cantidad considerable de valores ausentes, por lo que no sería del todo acertado el tomar estos resultados como certeros.

Con los datos completos ¿podría resultar ser el lunes como el día que más se escuchó de música?


